plugin/route53/route53.go

```diff
@@ -15,98 +24,275 @@ import (
-	})
-	if err != nil {
-		return dns.RcodeServerFailure, err
+	zones := make(map[string]*zone, len(keys))
+	zoneNames := make([]string, 0, len(keys))
+	for dns, id := range keys {
+		_, err := c.ListHostedZonesByNameWithContext(ctx, &route53.ListHostedZonesByNameInput{
+			DNSName:      aws.String(dns),
+			HostedZoneId: aws.String(id),
+		})
+		if err != nil {
+			return nil, err
+		}
+		zones[dns] = &zone{id: id, z: file.NewZone(dns, "")}
+		zoneNames = append(zoneNames, dns)
 	}
+	return &Route53{
+		client:    c,
+		zoneNames: zoneNames,
+		zones:     zones,
+		upstream:  up,
+	}, nil
+}
 
-	answers := []dns.RR{}
-	switch state.QType() {
-	case dns.TypeA:
-		answers = a(qname, output.ResourceRecordSets)
-	case dns.TypeAAAA:
-		answers = aaaa(qname, output.ResourceRecordSets)
-	case dns.TypePTR:
-		answers = ptr(qname, output.ResourceRecordSets)
+// Run executes first update, spins up an update forever-loop.
+// Returns error if first update fails.
+func (h *Route53) Run(ctx context.Context) error {
+	if err := h.updateZones(ctx); err != nil {
+		return err
 	}
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				log.Infof("breaking out of Route53 update loop: %v", ctx.Err())
+				return
+			case <-time.After(1 * time.Minute):
+				if err := h.updateZones(ctx); err != nil && ctx.Err() == nil /* Don't log error if ctx expired. */ {
+					log.Errorf("failed to update zones: %v", err)
+				}
+			}
+		}
+	}()
+	return nil
+}
 
-	if len(answers) == 0 {
-		return plugin.NextOrFailure(rr.Name(), rr.Next, ctx, w, r)
+// ServeDNS implements the plugin.Handler.ServeDNS.
+func (h *Route53) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r}
+	qname := state.Name()
+
+	zName := plugin.Zones(h.zoneNames).Matches(qname)
+	if zName == "" {
+		return plugin.NextOrFailure(h.Name(), h.Next, ctx, w, r)
+	}
+	z, ok := h.zones[zName]
+	if !ok || z == nil {
+		return dns.RcodeServerFailure, nil
 	}
 
 	m := new(dns.Msg)
 	m.SetReply(r)
 	m.Authoritative, m.RecursionAvailable = true, true
-	m.Answer = answers
+	var result file.Result
+	h.zMu.RLock()
+	m.Answer, m.Ns, m.Extra, result = z.z.Lookup(state, qname)
+	h.zMu.RUnlock()
+
+	switch result {
+	case file.Success:
+	case file.NoData:
+	case file.NameError:
+		m.Rcode = dns.RcodeNameError
+	case file.Delegation:
+		m.Authoritative = false
+	case file.ServerFailure:
+		return dns.RcodeServerFailure, nil
+	}
 
 	state.SizeAndDo(m)
 	m, _ = state.Scrub(m)
 	w.WriteMsg(m)
 	return dns.RcodeSuccess, nil
 }
 
-func a(zone string, rrss []*route53.ResourceRecordSet) []dns.RR {
-	answers := []dns.RR{}
-	for _, rrs := range rrss {
-		for _, rr := range rrs.ResourceRecords {
-			r := new(dns.A)
-			r.Hdr = dns.RR_Header{Name: zone, Rrtype: dns.TypeA, Class: dns.ClassINET, Ttl: uint32(aws.Int64Value(rrs.TTL))}
-			r.A = net.ParseIP(aws.StringValue(rr.Value)).To4()
-			answers = append(answers, r)
+func awsStrTypeToUint16(t string) (uint16, bool) {
```

Comment:
this can be replaced with `dns.TypeToString[]` lookups. Prolly still need a switch to filter on the knows records.